src/conventional/amdar_bufr2ioda.py

```diff
@@ -0,0 +1,529 @@
+    return False
+
+
+def read_file(file_name, count, start_pos, data):
+
+    f = open(file_name, 'rb')
+
+    while True:
+        # Use eccodes to decode each bufr message in the file
+        data, count, start_pos = read_bufr_message(f, count, start_pos, data)
+
+        if start_pos is None:
+            break
+
+    return data, count, start_pos
+
+
+def read_bufr_message(f, count, start_pos, data):
+
+    meta_data = {}         # All the various MetaData go in here.
+    vals = {}              # Floating-point ObsValues variables go in here.
+    avals = []             # Temporarily hold array of values.
+    call_fail = False
+    if start_pos == f.tell():
+        return data, count, None
+    start_pos = f.tell()
+    count[0] += 1
+    logging.info("BUFR message number: " + str(count[0]))
+
+    met_utils = meteo_utils.meteo_utils()
+
+    try:
+        bufr = ecc.codes_bufr_new_from_file(f)
+    except:  # noqa
+        logging.critical("ABORT, failue when attempting to call:  codes_bufr_new_from_file")
+
+    try:
+        ecc.codes_set(bufr, 'skipExtraKeyAttributes', 1)   # Supposedly this is ~25 percent faster
+        ecc.codes_set(bufr, 'unpack', 1)
+    except:  # noqa
+        logging.info("finished unpacking BUFR file")
+        start_pos = None
+        return data, count, start_pos
+
+    # First, get the MetaData we are interested in (list is in metaDataKeyList)
+    for k, v in metaDataKeyList.items():
+        meta_data[k] = []
+        if (len(v) > 1):
+            for var in v:
+                if (var != 'Constructed'):
+                    try:
+                        avals = ecc.codes_get_array(bufr, var)
+                        meta_data[k] = assign_values(avals, k)
+                        if not is_all_missing(meta_data[k]):
+                            break
+                    except ecc.KeyValueNotFoundError:
+                        logging.warning("Caution: unable to find requested BUFR key: " + var)
+        else:
+            if (v[0] != 'Constructed'):
+                try:
+                    avals = ecc.codes_get_array(bufr, v[0])
+                    meta_data[k] = assign_values(avals, k)
+                except ecc.KeyValueNotFoundError:
+                    logging.warning("Caution, unable to find requested BUFR key: " + v[0])
+
+    # Determine the target number of observation points from a critical variable (i.e., latitude).
+    target_number = len(meta_data[var_mimic_length])
+
+    # For any of the MetaData elements that were totally lacking, fill entire vector with missing.
+    empty = []
+    for k, v in metaDataKeyList.items():
+        if not any(meta_data[k]):
+            meta_data[k] = assign_missing_meta(empty, k, target_number, 0)
+        if (len(meta_data[k]) == 1):
+            meta_data[k] = np.full(target_number, meta_data[k][0])
+        elif (len(meta_data[k]) < target_number):
+            logging.warning("Key called " + k + " contains only " + str(len(meta_data[k])) +
```

Comment:
```suggestion
            logging.warning(f"Key called {k} contains only {len(meta_data[k])} "
                                        f" elements, wheras {target_number} were expected.")
```

Reply:
above should work?

Reply:
Very good suggestion, thanks @CoryMartin-NOAA.  I find it really weird syntax with this f character to precede something in double-quotes and so forth.

Reply:
https://docs.python.org/3/tutorial/inputoutput.html#tut-f-strings formatted string literals